### ISSUE TITLE
[Request] Is there any possability of adding in addition to Python, the ability to trigger PowerShell scripts from hotkeys?

### ISSUE THREAD

**Issue by MostHated** [autokey/autokey#513]
Hey there,
    I don't have an issue per se, but was wondering if it might be possible to add the ability to trigger/call a [PowerShell](https://docs.microsoft.com/en-us/powershell/scripting/install/installing-powershell-core-on-linux?view=powershell-7.1) script as a secondary option to Python? If not, I can certainly understand but figured it was worth an ask.

Thanks,
-MH

**Comment by josephj11** [comment]
That's a new one! Didn't know you could run PowerShell on Linux.

TL;DR: You have to learn a tiny bit of Python 3. It will let you do almost anything.

The answer is easy. As far as Linux is concerned, PowerShell is just another program, so you just tell it to run it.

Since AutoKey scripts are written in Python 3, you have to use that to talk to Linux. For what you want, the subprocess module is what you're looking for.

Before you can use it, you have to make it available to your script with

    import subprocess

There are several API calls depending on what you want the called program to return to your AutoKey Python script. 

If your program doesn't return any output (on stdout) that you want, use this

    subprocess.Popen(["path to your program", "first argument to your program", ... ])

or

    subprocess.call(["path to your program", "first argument to your program", ... ])

Use Popen if you want the next step of your Python script to run immediately. Use Call if you want to wait until your program terminates before running the next step of your Python script.

This one returns the output (from stdout) of your program in a Python variable (result).

    result = subprocess.check_output(["path to your program", "first argument to your program", ... ])

A few notes:
Python is a large and powerful language with modules available to do almost anything, so the above examples just scratch the surface.

If any of your arguments are variables instead of strings, then don't put them in quotes.
If there is any possibility that your external command can fail or return an error status, then call it within a try: ... except block.

All this being said, why would you want to do this? Python and shells like bash are far more powerful and native to Linux, so they'll also have better support.

----

BTW: Whenever you have a question (as opposed to an issue) it's better to post it on our [list](https://groups.google.com/forum/#!forum/autokey-users) or on [Gitter](https://gitter.im/autokey/autokey). The list is more user oriented whereas more technical questions will get a better answer on Gitter.

Both of these venues have a number of active users, so you are likely to get more, better, and faster answers on them than here on GitHub where only our very busy developers and a few users see the posts. We also prefer to have the developers developing rather than taking time to answer questions and manage issues that the user community could handle for them.

**Comment by MostHated** [comment]
My apologies for posting this to the issue board, but I appreciate the detailed reply. It seems to be more rare than it should be for a project to have a dedicated and separate support forum, so I didn't think to check for that. I ended up going the route you mentioned above and things are working well, so thank you much for that. 

A side note, I know it is somewhat blasphemous, but I rather like having PowerShell on Linux as I already have dozens of scripts written using it, and when using built-in environmental variables, such as $HOME, my scripts all work cross-platform. Mix that with being able to incorporate .Net libraries and C# code on the fly, I find myself writing more and more of my Linux scripts with it.

<details>
<summary>Runs just fine on Linux and Windows as is straight from the terminal!</summary>

```PowerShell
# -- helloworld.ps1

$CSharpCode = @'
using System;
namespace HelloWorld
{
    public class Program
    {
        public static void Main(){
            Console.WriteLine("Hello world!");
        }
    }
}
'@

Add-Type -TypeDefinition $CSharpCode -Language CSharp
[HelloWorld.Program]::Main() 
```

</details>

That said, I made sure to join and bookmark the Google group incase I end up having any more questions in the future.
Thanks!
-MH
